README.txt
= RightScale Amazon Web Services Ruby Gems

Published by RightScale, Inc. under the MIT License.
For information about RightScale, see http://www.rightscale.com
Maintained by the RightScale “Orange_team”

 Maintained by the RightScale "Orange_team"

== DESCRIPTION:

The RightScale AWS gems have been designed to provide a robust, fast, and secure interface to Amazon EC2, EBS, S3, SQS, SDB, and CloudFront. 
These gems have been used in production by RightScale since late 2006 and are being maintained to track enhancements made by Amazon. 
The RightScale AWS gems comprise:

- RightAws::Ec2 -- interface to Amazon EC2 (Elastic Compute Cloud), VPC (Virtual Private Cloud) and the associated EBS (Elastic Block Store)
- RightAws::S3 and RightAws::S3Interface -- interface to Amazon S3 (Simple Storage Service)
- RightAws::Sqs and RightAws::SqsInterface -- interface to first-generation Amazon SQS (Simple Queue Service)
- RightAws::SqsGen2 and RightAws::SqsGen2Interface -- interface to second-generation Amazon SQS (Simple Queue Service)
- RightAws::SdbInterface and RightAws::ActiveSdb -- interface to Amazon SDB (SimpleDB)
- RightAws::AcfInterface -- interface to Amazon CloudFront, a content distribution service
- RightAws::AsInterface  -- interface to Amazon Auto Scaling
- RightAws::AcwInterface -- interface to Amazon Cloud Watch
- RightAws::ElbInterface -- interface to Amazon Elastic Load Balancer
- RightAws::RdsInterface -- interface to Amazon RDS instances

== FEATURES:

- Full programmmatic access to EC2, EBS, S3, SQS, SDB, CloudFront, AS, ACW, ELB and RDS.
- Complete error handling: all operations check for errors and report complete
  error information by raising an AwsError.
- Persistent HTTP connections with robust network-level retry layer using
  RightHttpConnection).  This includes socket timeouts and retries.
- Robust HTTP-level retry layer.  Certain (user-adjustable) HTTP errors returned
  by Amazon's services are classified as temporary errors.
  These errors are automaticallly retried using exponentially increasing intervals.
  The number of retries is user-configurable.
- Fast REXML-based parsing of responses (as fast as a pure Ruby solution allows).
- Uses libxml (if available) for faster response parsing. 
- Support for large S3 list operations.  Buckets and key subfolders containing
  many (> 1000) keys are listed in entirety.  Operations based on list (like
  bucket clear) work on arbitrary numbers of keys.
- Support for streaming GETs from S3, and streaming PUTs to S3 if the data source is a file.
- Support for single-threaded usage, multithreaded usage, as well as usage with multiple
  AWS accounts.
- Support for both first- and second-generation SQS (API versions 2007-05-01
  and 2008-01-01).  These versions of SQS are not compatible.
- Support for signature versions 0 and 1 on SQS, SDB, and EC2.
- Interoperability with any cloud running Eucalyptus (http://eucalyptus.cs.ucsb.edu)
- Test suite (requires AWS account to do "live" testing).

== THREADING:

All RightScale AWS interfaces offer two threading options:
1. Use a single persistent HTTP connection per process.
2. Use a persistent HTTP connection per Ruby thread.
 
Either way, it doesn't matter how many (for example) RightAws::S3 objects you create,
they all use the same per-program or per-thread
connection. The purpose of sharing the connection is to keep a single
persistent HTTP connection open to avoid paying connection
overhead on every request. However, if you have multiple concurrent
threads, you may want or need an HTTP connection per thread to enable
concurrent requests to AWS. The way this plays out in practice is:
1. If you have a non-multithreaded Ruby program, use the non-multithreaded setting.
2. If you have a multi-threaded Ruby program, use the multithreaded setting to enable
   concurrent requests to S3 (or SQS, or SDB, or EC2).
3. For running under Mongrel/Rails, use the non-multithreaded setting even though
   mongrel is multithreaded.  This is because only one Rails handler is invoked at
   time (i.e. it acts like a single-threaded program)

Note that due to limitations in the I/O of the Ruby interpreter you
may not get the degree of parallelism you may expect with the multi-threaded setting.

== GETTING STARTED:

* For EC2 read RightAws::Ec2 and consult the Amazon EC2 API documentation at
  http://developer.amazonwebservices.com/connect/kbcategory.jspa?categoryID=87
* For S3 read RightAws::S3 and consult the Amazon S3 API documentation at
  http://developer.amazonwebservices.com/connect/kbcategory.jspa?categoryID=48
* For first generation SQS read RightAws::Sqs and consult the Amazon SQS API documentation at
  http://developer.amazonwebservices.com/connect/kbcategory.jspa?categoryID=31
* For second generation SQS read RightAws::SqsGen2, RightAws::SqsGen2Interface, and consult the Amazon SQS API documentation at
  http://developer.amazonwebservices.com/connect/entry.jspa?externalID=1214&categoryID=31 

  Amazon's Migration Guide for moving from first to second generation SQS is
  avalable at:
  http://developer.amazonwebservices.com/connect/entry.jspa?externalID=1148
* For SDB read RightAws::SdbInterface, RightAws::ActiveSdb, and consult the Amazon SDB API documentation at
  http://developer.amazonwebservices.com/connect/kbcategory.jspa?categoryID=141
* For CloudFront (ACF) read RightAws::AcfInterface and consult the Amazon CloudFront API documentation at 
  http://developer.amazonwebservices.com/connect/kbcategory.jspa?categoryID=213

== KNOWN ISSUES:

- 7/08: A user has reported that uploads of large files on Windows may be broken on some
  Win platforms due to a buggy File.lstat.size.  Use the following monkey-patch at your own risk, 
  as it has been proven to break Rails 2.0 on Windows:

    require 'win32/file'
    class File
      def lstat
        self.stat
      end
    end


- Attempting to use the Gibberish plugin (used by the Beast forum app) 
  will break right_aws as well as lots of other code.  Gibberish
  changes the semantics of core Ruby (specifically, the String class) and thus presents a reliability
  problem for most Ruby programs.

- 2/11/08: If you use RightAws in conjunction with attachment_fu, the
  right_aws gem must be included (using the require statement) AFTER
  attachment_fu.  If right_aws is loaded before attachment_fu, you'll
  encounter errors similar to:

  s3.amazonaws.com temporarily unavailable: (wrong number of arguments (5 for 4))

  or

  'incompatible Net::HTTP monkey-patch'

  This is due to a conflict between the right_http_connection gem and another
  gem required by attachment_fu.

- 8/07: Amazon has changed the semantics of the SQS service.  A
  new queue may not be created within 60 seconds of the destruction of any
  older queue with the same name.  Certain methods of RightAws::Sqs and
  RightAws::SqsInterface will fail with the message:
  "AWS.SimpleQueueService.QueueDeletedRecently: You must wait 60 seconds after deleting a queue before you can create another with the same name."
  
== REQUIREMENTS:

RightAws requires REXML and the right_http_connection gem.
If libxml and its Ruby bindings (distributed in the libxml-ruby gem) are
present, RightAws can be configured to use them:
  RightAws::RightAWSParser.xml_lib = 'libxml'
Any error with the libxml installation will result in RightAws failing-safe to
REXML parsing.

== INSTALL:

sudo gem install right_aws

== LICENSE:

Copyright (c) 2007-2013 RightScale, Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.1.0 2007-08-10
Initial release.

== 1.2.0 2007-09-12

* r1718, todd, 2007-09-12 15:34:37
  * # 458, Extensive documentation review, rework, and expansion.  Also added
    coverage analysis to the test suite using RCov.

* r1690, todd, 2007-09-07 15:23:11
  * # 447, Add support.rb to manifest

* r1688, todd, 2007-09-07 13:57:39
  * # 447, Use Active Support if available, but don't require it.  Load our own
    extensions if it's not present.  This keeps us from overloading ActiveSupport if
    a user's already using it.

* r1687, todd, 2007-09-07 11:36:43
  * # 447, Removed dependency on activesupport

* r1676, konstantin, 2007-09-06 01:27:09
  * paid AMIs, small fix

* r1667, konstantin, 2007-09-05 12:58:10
  * # 427, paid AMI support for ec2_instances

* r1658, konstantin, 2007-09-05 01:02:25
  * params improvements for ec2.describe_xxx, now ones can use a String as well as an Array.

* r1653, konstantin, 2007-09-04 12:31:19
  * libxml and paid AMI support added

* r1581, tve, 2007-08-24 16:21:45
  * Improved RightAws documentation

== 1.3.0 2007-09-26

* r1754, todd, 2007-09-19 13:48:34
  * # 487, # 488, Consolidate a lot of code that was repeated in three places.
    Fix error handling path when using streaming GET interfaces with S3.
    Also add a stub for RightHttpConnection which allows more control over the
    unit tests.  Expand the unit tests and coverage tests.

* r1755, todd, 2007-09-19 14:29:19
  * # 487, RDoc fixes after code consolidation

* r1866, konstantin, 2007-10-05 06:17:36
  * # 220, close connection on HTTP 5xx/4xx errors

== 1.4.0 2007-10-10

* r1868, konstantin, 2007-10-05 06:38:37
  * # 220, inst_type branch merge into 1.3.0 release

* r1869, konstantin, 2007-10-05 07:32:34
  * right_http_connection 1.1.1 requirements fixed

* r1879, konstantin, 2007-10-07 02:11:24
  * # 524, blocks added to ec2_describe_xxx to support aws_cache

* r1924, konstantin, 2007-10-12 11:35:06
  * # 536, user_data bug fix

* r1929, tve,  2007-10-15 00:00:11
  * Fix libxml/rexml selection bug

* r1938, konstantin, 2007-10-16 10:53:56
  * instance type support is set to default

== 1.4.3 2007-10-25

* r1983, konstantin, 2007-10-25 22:33:00 +0400
  * Fixed ActiveSupport requirement bug (thanks to Toby)
  * Fixed HttpConnection logging to stdout bug (thanks to Toby)

== 1.4.4

* r1999, tve, 2007-11-01 00:07:00 -0700
  * Fixed escaping issue affecting key names in S3 gem
  * Fixed duplicate marker in S3 incremental bucket listing

* r2001, konstantin, 2007-11-01 12:03:13  +0300
  * Fixed multiple permissions assignment on Grantee#grant/revoke
  * Fixed new grantee permissions set ingnore (Grantee#apply)
  * S3::Grantee#exists? method added

* r2109, konstantin, 2007-11-12 21:49:36  +0300
  * RightAwsBaseInterface: caching implemented.
    (The Ec2 functions are being cached: describe_images, describe_instances,
    describe_security_groups and describe_key_pairs)

== 1.4.5 - 1.4.6
* r 2619, konstantin, 01-17-08 16:18:36  +0300
  * S3 Location constraints support added.
  * Fixed bug with trailing '/' in the bucket name for 'EU' located buckets
  * Added: S3Interface#bucket_location, S3::Bucket#location

== 1.4.7
* r 2622, konstantin, 01-18-08 13:52:20  +0300
  * Virtual domains doc added
  * S3 Query API fixed to support virtual domains.

== 1.4.8
* r 2650, konstantin, 01-24-08 11:12:00  +0300
  * net_fix.rb moved to right_http_connection

== 1.5.0
* r 2688, konstantin, 02-30-08 15:42:00  +0300
  * SDB support added.
  * RightAws::S3::bucket and RightAws::S3::Bucket.create methods behaviour
    changed: param +create+ is set to +false+ by default.

== 1.6.0

* r2780, todd, 2008-02-11 11:41:07 -0800 (Mon, 11 Feb 2008), 4 lines
  * Some doc updates & tweaks: we now support ultra-large PUTs to S3, small SDB
    cleanups, add SDB to the list of interfaces in RightAws, update some copyrights,
    warn about loading attachment_fu AFTER right_aws (big no-no).

* r2784, todd, 2008-02-11 13:46:47 -0800 (Mon, 11 Feb 2008), 2 lines
  * One final clarification: you may get a Net::HTTP bad monkey patch exception

* r2880, todd, 2008-02-25 18:06:22 -0800 (Mon, 25 Feb 2008), 2 lines
  * Add SQS 'Gen 2' interface implementation

* r2913, todd, 2008-02-29 16:57:07 -0800 (Fri, 29 Feb 2008), 3 lines
  * SqsGen2 (object interface), unit tests for both interfaces, documentation updates.

* r2922, todd, 2008-03-03 15:26:42 -0800 (Mon, 03 Mar 2008), 2 lines
  * couple of documentation tweaks in prep for 1.6.0

== 1.6.1

* r2963, todd, 2008-03-06 19:10:23 -0800 (Thu, 06 Mar 2008), 3 lines
  * (#950) Many minor fixes in incrementally_list_bucket to prevent a death loop
    in certain rare conditions

== 1.7.0

* r3051, konstantin, 2008-03-14 21:26:12 +0300 (Fri, 14 Mar 2008), 1 line
  * #897, ActiveSdb alpha release

== 1.7.1

  Do not autoload right_sdb with the rest of the modules; it requires uuidtools.
    We want the user to explicly request ActiveSdb, at least as long as it is
    alpha/beta.

  Fix escaping problem in SqsGen2Interface: POST bodies did not properly escape the '&' character

== 1.7.2

  Release Notes:

  RightAws includes some new features, including:
  - Support in RightAws::S3 and RightAws::S3Interface for S3 key copy, move, and rename
  - Support for signature version 0 request authentication to EC2, SQS, and SDB
  - Enhanced S3 object meta-header read and update
  - Interoperability with clouds running Eucalyptus (http://eucalyptus.cs.ucsb.edu)
    [ Contributed by the Eucalyptus group ]
  - Support for c1.medium and c1.xlarge instance types

  Bug fixes include:
  - Corrected the failure, under certain conditions, of retries of streaming PUTs to S3.
    We now reset the seek pointer of the streaming IO object to its initial position.
  - Removal of an accidental dependency on ActiveSupport in RightAws::S3Interface.get_link().
  - Monkey-patch of the Ruby File class on Windows platforms to correct a problem in lstat.

lib/lbs/right_lbs.rb
#
# Copyright (c) 2007-2009 RightScale Inc
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

module RightAws

  # Load Balancing Service
  class Lbs < RightAwsBase
    include RightAwsBaseInterface

    # Amazon LBS API version being used
    API_VERSION       = "2009-01-22"
    DEFAULT_HOST      = "lbs.amazonaws.com"
    DEFAULT_PATH      = '/'
    # KD: FIXME later to https and 443
    DEFAULT_PROTOCOL  = 'http'
    DEFAULT_PORT      = 80

    @@bench = AwsBenchmarkingBlock.new
    def self.bench_xml
      @@bench.xml
    end
    def self.bench_service
      @@bench.service
    end

    # Create a new handle to an LBS account. All handles share the same per process or per thread
    # HTTP connection to Amazon LBS. Each handle is for a specific account. The params have the
    # following options:
    # * <tt>:endpoint_url</tt> a fully qualified url to Amazon API endpoint (this overwrites: :server, :port, :service, :protocol and :region). Example: 'https://eu-west-1.LBS.amazonaws.com/'
    # * <tt>:server</tt>: LBS service host, default: DEFAULT_HOST
    # * <tt>:region</tt>: LBS region (North America by default)
    # * <tt>:port</tt>: LBS service port, default: DEFAULT_PORT
    # * <tt>:protocol</tt>: 'http' or 'https', default: DEFAULT_PROTOCOL
    # * <tt>:multi_thread</tt>: true=HTTP connection per thread, false=per process
    # * <tt>:logger</tt>: for log messages, default: RAILS_DEFAULT_LOGGER else STDOUT
    # * <tt>:signature_version</tt>:  The signature version : '0','1' or '2'(default)
    # * <tt>:cache</tt>: true/false(default): caching works for: describe_access_points
    #
    def initialize(aws_access_key_id=nil, aws_secret_access_key=nil, params={})
      init({ :name                => 'LBS',
             :default_host        => ENV['LBS_URL'] ? URI.parse(ENV['LBS_URL']).host   : DEFAULT_HOST,
             :default_port        => ENV['LBS_URL'] ? URI.parse(ENV['LBS_URL']).port   : DEFAULT_PORT,
             :default_service     => ENV['LBS_URL'] ? URI.parse(ENV['LBS_URL']).path   : DEFAULT_PATH,
             :default_protocol    => ENV['LBS_URL'] ? URI.parse(ENV['LBS_URL']).scheme : DEFAULT_PROTOCOL,
             :default_api_version => ENV['LBS_API_VERSION'] || API_VERSION },
           aws_access_key_id    || ENV['AWS_ACCESS_KEY_ID'] ,
           aws_secret_access_key|| ENV['AWS_SECRET_ACCESS_KEY'],
           params)
    end

    def generate_request(action, params={}) #:nodoc:
      generate_request_impl(:get, action, params )
    end

      # Sends request to Amazon and parses the response
      # Raises AwsError if any banana happened
    def request_info(request, parser)  #:nodoc:
      request_info_impl(:lbs_connection, @@bench, request, parser)
    end

    #-----------------------------------------------------------------
    #      Access Points
    #-----------------------------------------------------------------

    # Describe access points.
    # Returns an array of access points.
    #
    #   lbs.describe_access_points #=>
    #    [{:dns_name     => "konstantin-5626.us-east.lbs.amazonaws.com",
    #      :health_check =>
    #       {:ping_target         => "TCP:90",
    #        :timeout             => 4,
    #        :healthy_threshold   => 5,
    #        :unhealthy_threshold => 5,
    #        :interval            => 5},
    #      :availability_zones => ["us-east-1c"],
    #      :routing_protocol   => "HTTP",
    #      :access_point_port  => 80,
    #      :end_points         => ["i-b070fed9"],
    #      :end_point_port     => 80,
    #      :created_time       => Tue Feb 03 20:35:51 UTC 2009,
    #      :access_point_name  => "konstantin"},
    #     {:dns_name     => "test-5643.us-east.lbs.amazonaws.com",
    #      :health_check =>
    #       {:ping_target         => "TCP:80",
    #        :timeout             => 5,
    #        :healthy_threshold   => 10,
    #        :unhealthy_threshold => 2,
    #        :interval            => 30},
    #      :availability_zones => ["us-east-1c", "us-east-1a"],
    #      :routing_protocol   => "HTTP",
    #      :access_point_port  => 80,
    #      :end_points         => [],
    #      :end_point_port     => 80,
    #      :created_time       => Wed Feb 04 09:19:18 UTC 2009,
    #      :access_point_name  => "test"}]
    #
    def describe_access_points(*access_points)
      access_points = access_points.flatten.compact
      request_hash = amazonize_list("AccessPointNames.member", access_points)
      link = generate_request("DescribeAccessPoints", request_hash)
      request_cache_or_info(:describe_access_points, link,  DescribeAccessPointsParser, @@bench, access_points.blank?)
    end

    # Create new access point.
    # Returns a new access point DNS name.
    #
    #  +Options+ is a hash:
    #    { :routing_protocol  => string  (:http(default) | :https | :tcp),
    #      :access_point_port => integer (80(default), 443 or 1024..65535),
    #      :end_point_port    => integer (1..65535) }
    #
    #  lbs.create_access_point('test', ['us-east-1a','us-east-1c']) #=> "test-5643.us-east.lbs.amazonaws.com"
    #
    def create_access_point(access_point_name, availability_zones=[], options={})
      request_hash = amazonize_list("AvailabilityZones.member", availability_zones)
      request_hash.merge!( 'AccessPointName'   =>  access_point_name,
                           'RoutingProtocol'   => (options[:routing_protocol]  || :http).to_s.upcase,
                           'AccessPointPort'   =>  options[:access_point_port] || 80,
                           'EndPointPort'      =>  options[:end_point_port]    || 80 )
      link = generate_request("CreateAccessPoint", request_hash)
      request_info(link, CreateAccesspointParser.new(:logger => @logger))
    end

    # Delete access point.
    # Returns +true+ on success.
    def delete_access_point(access_point_name)
      link = generate_request("DeleteAccessPoint", 'AccessPointName' => access_point_name )
      request_info(link, DeleteAccesspointParser.new(:logger => @logger))
    end

    # Add one or more zones to an access point.
    # Returns a list of updated availability zones for the access point.
    #
    #  lbs.add_availability_zones("test", "us-east-1c") #=> ["us-east-1a", "us-east-1c"]
    #
    def add_availability_zones(access_point_name, *availability_zones)
      availability_zones.flatten!
      request_hash = amazonize_list("AvailabilityZones.member", availability_zones)
      request_hash.merge!( 'AccessPointName' => access_point_name )
      link = generate_request("AddAvailabilityZones", request_hash)
      request_info(link, AvailabilityZonesParser.new(:logger => @logger))
    end

    # Remove one or more zones from an access point.
    # Returns a list of updated availability zones for the access point.
    #
    #  lbs.add_availability_zones("test", "us-east-1c") #=> ["us-east-1a"]
    #
    def remove_availability_zones(access_point_name, *availability_zones)
      availability_zones.flatten!
      request_hash = amazonize_list("AvailabilityZones.member", availability_zones)
      request_hash.merge!( 'AccessPointName' => access_point_name )
      link = generate_request("RemoveAvailabilityZones", request_hash)
      request_info(link, AvailabilityZonesParser.new(:logger => @logger))
    end

    # Configure an access points health check.
    # Returns an updated health check configuration for the access point.
    #
    #  +Health_check+ is a hash:
    #    { :interval            => integer,
    #      :ping_target         => string,
    #      :timeout             => integer,
    #      :healthy_threshold   => integer,
    #      :unhealthy_threshold => integer }
    #
    def configure_health_check(access_point_name, health_check)
      request_hash = { 'AccessPointName' => access_point_name }
      health_check.each{ |key, value| request_hash["HealthCheck.#{key.to_s.camelize}"] = value }
      link = generate_request("ConfigureHealthCheck", request_hash)
      request_info(link, HealthCheckParser.new(:logger => @logger))
    end

    #-----------------------------------------------------------------
    #      End Points
    #-----------------------------------------------------------------

    # Describe end points state.
    # Returns a list of end point states containing InService or OutOfService
    # and any reasons associated with the OutOfService state.
    #
    # lbs.describe_end_point_state('test') #=>
    #    [{:instance_id => "i-b070fed9",
    #      :state       => "OutOfService",
    #      :description => "EndPoint has failed at least the UnhealthyThreshold number of health checks consecutively.",
    #      :reason_code => "EndPoint"},
    #     {:instance_id => "i-73149a1a",
    #      :state       => "InService",
    #      :description => "N/A",
    #      :reason_code => "N/A"}]
    #
    def describe_end_point_state(access_point_name, *end_points)
      end_points.flatten!
      request_hash = amazonize_list("EndPoints.member.?.InstanceId", end_points)
      request_hash.merge!( 'AccessPointName' => access_point_name )
      link = generate_request("DescribeEndPointState", request_hash)
      request_info(link, DescribeEndPointStateParser.new(:logger => @logger))
    end

    # Add new instance(s) to the access point.
    # Returns an updated list of endpoints for the access point.
    #
    #  lbs.register_end_points('test', "i-b070fed9") #=> ["i-b070fed9"]
    #
    def register_end_points(access_point_name, *end_points)
      end_points.flatten!
      request_hash = amazonize_list("EndPoints.member.?.InstanceId", end_points)
      request_hash.merge!( 'AccessPointName' => access_point_name )
      link = generate_request("RegisterEndPoints", request_hash)
      request_cache_or_info :register_end_points, link, EndPointsParser, @@bench, end_points.blank?
    end

    # Remove instance(s) from the access point.
    # Returns an updated list of endpoints for the access point.
    #
    #  lbs.deregister_end_points('test', "i-b070fed9") #=> []
    #
    def deregister_end_points(access_point_name, *end_points)
      end_points.flatten!
      request_hash = amazonize_list("EndPoints.member.?.InstanceId", end_points)
      request_hash.merge!( 'AccessPointName' => access_point_name )
      link = generate_request("DeregisterEndPoints", request_hash)
      request_cache_or_info :deregister_end_points, link, EndPointsParser, @@bench, end_points.blank?
    end

    #-----------------------------------------------------------------
    #      PARSERS: Access Points
    #-----------------------------------------------------------------
 
    class DescribeAccessPointsParser < RightAWSParser #:nodoc:
      def tagstart(name, attributes)
        case name
        when 'member'
          case @xmlpath
            when 'DescribeAccessPointsResponse/DescribeAccessPointsResult/AccessPointsDescriptions'
              @item = { :availability_zones => [],
                        :health_check       => {},
                        :end_points         => [] }
          end
        end
      end
      def tagend(name)
        case name
        when 'AccessPointName'    then @item[:access_point_name] = @text
        when 'AccessPointPort'    then @item[:access_point_port] = @text.to_i
        when 'DNSName'            then @item[:dns_name]          = @text
        when 'EndPointPort'       then @item[:end_point_port]    = @text.to_i
        when 'CreatedTime'        then @item[:created_time]      = Time::parse(@text)
        when 'RoutingProtocol'    then @item[:routing_protocol]  = @text
        when 'InstanceId'         then @item[:end_points]        << @text
        when 'Interval'           then @item[:health_check][:interval]            = @text.to_i
        when 'PingTarget'         then @item[:health_check][:ping_target]         = @text
        when 'HealthyThreshold'   then @item[:health_check][:healthy_threshold]   = @text.to_i
        when 'Timeout'            then @item[:health_check][:timeout]             = @text.to_i
        when 'UnhealthyThreshold' then @item[:health_check][:unhealthy_threshold] = @text.to_i
        when 'member'
          case @xmlpath
          when 'DescribeAccessPointsResponse/DescribeAccessPointsResult/AccessPointsDescriptions'
            @result << @item
          when 'DescribeAccessPointsResponse/DescribeAccessPointsResult/AccessPointsDescriptions/member/AvailabilityZones'
            @item[:availability_zones] << @text
          end
        end
      end
      def reset
        @result = []
      end
    end

    class CreateAccesspointParser < RightAWSParser #:nodoc:
      def tagend(name)
        @result = @text if name == 'DNSName'
      end
    end

    class DeleteAccesspointParser < RightAWSParser #:nodoc:
      def tagend(name)
        @result = true if name == 'DeleteAccessPointResult'
      end
    end

    class AvailabilityZonesParser < RightAWSParser #:nodoc:
      def tagend(name)
        @result << @text if name == 'member'
      end
      def reset
        @result = []
      end
    end

    class HealthCheckParser < RightAWSParser #:nodoc:
      def tagend(name)
        case name
        when 'Interval'           then @result[:interval]            = @text.to_i
        when 'PingTarget'         then @result[:ping_target]         = @text
        when 'HealthyThreshold'   then @result[:healthy_threshold]   = @text.to_i
        when 'Timeout'            then @result[:timeout]             = @text.to_i
        when 'UnhealthyThreshold' then @result[:unhealthy_threshold] = @text.to_i
        end
      end
      def reset
        @result = {}
      end
    end

    #-----------------------------------------------------------------
    #      PARSERS: End Points
    #-----------------------------------------------------------------

    class DescribeEndPointStateParser < RightAWSParser #:nodoc:
      def tagstart(name, attributes)
        @item = {} if name == 'member'
      end
      def tagend(name)
        case name
        when 'Description' then @item[:description] = @text
        when 'State'       then @item[:state]       = @text
        when 'InstanceId'  then @item[:instance_id] = @text
        when 'ReasonCode'  then @item[:reason_code] = @text
        when 'member'      then @result            << @item
        end
      end
      def reset
        @result = []
      end
    end

    class EndPointsParser < RightAWSParser #:nodoc:
      def tagend(name)
        @result << @text if name == 'InstanceId'
      end
      def reset
        @result = []
      end
    end

  end

end
